test/distributed/fsdp/test_fsdp.py

```diff
@@ -0,0 +1,259 @@
+
+        self.run_test(test_name, pipe)
+
+        dist.barrier()
+
+        dist.destroy_process_group()
+        sys.exit(0)
+
+    def _train_for_several_steps(self, model, num_steps, autocast, lr=0.01, norm_type=None):
+        model_device = next(model.parameters()).device
+        # use SGD with momentum instead of Adam, since Adam is scale invariant
+        # and this makes it bad for tests
+        optim = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
+        for _ in range(num_steps):
+            optim.zero_grad()
+            with torch.cuda.amp.autocast(enabled=autocast):
+                # Inputs always cuda regardless of cpu offloading, or model.device
+                input = model.module.get_input(torch.device("cuda"))
+                output = model(*input)
+                loss = model.module.get_loss(input, output).to(model_device)
+            assert loss.dtype == torch.float32
+            model.module.run_backward(loss)
+            optim.step()
+        if isinstance(model, FullyShardedDataParallel):
+            model._assert_state(TrainingState.IDLE)
+        return loss.detach()
+
+    def _test_identical_outputs(
+        self, model_init_fn, num_steps=2, use_cuda=True, lr=0.01, ref_ddp_fn=None, norm_type=None,
+    ):
+        group = dist.distributed_c10d._get_default_group()
+        rank = group.rank()
+        # Establish reference behavior with PyTorch DDP (+ optionally autocast).
+        model = model_init_fn(group=group, wrap_fsdp=False).cuda()
+        if ref_ddp_fn is None:
+            model = nn.parallel.DistributedDataParallel(
+                model, device_ids=[rank], output_device=rank
+            )
+        else:
+            model = ref_ddp_fn(model, group)
+        ref_loss = self._train_for_several_steps(model, num_steps, autocast=False, lr=lr, norm_type=norm_type)
+        ref_full_params = list(model.parameters())
+
+        # Confirm we get the same behavior using FullyShardedDataParallel.
+        model = model_init_fn(group=group, wrap_fsdp=True)
+        model = FullyShardedDataParallel(model)
+        if use_cuda:
+            model = model.cuda()
+        else:
+            assert next(model.parameters()).device == torch.device("cpu")
+        shard_loss = self._train_for_several_steps(model, num_steps, autocast=False, lr=lr, norm_type=norm_type)
+        _get_full_params(model)
+        shard_full_params = list(model.parameters())
+
+        try:
+            torch.testing.assert_allclose(ref_loss, shard_loss)
+            self.assertEqual(ref_full_params, shard_full_params, exact_device=True)
+        except (AssertionError, RuntimeError) as e:
+            raise Exception(f"FullyShardedDataParallel didn't match PyTorch DDP\n\n {e}")
+
+class TestParityWithDDP(FSDPTest):
+    """
+    Compare losses and parameter values after several updates when using
+    PyTorch DDP vs. FullyShardedDataParallel.
+    """
+
+    @skip_if_lt_x_gpu(2)
+    @skip_if_rocm
```

Comment:
Why do you need these skips for rocm and windows? The runner should never run these files in these context right?